### PR TITLE
Add CustomResourceGenerator struct, generate metadata field correctly…

### PR DESCRIPTION
**Pull request by albert-zhong**
Added a bunch of miscellaneous fixes. I've been testing `crd2pulumi` on some larger, ~5k line CRDs in cert-manager and they all seem to correctly generate. I'm sure there's some edge cases with the OpenAPI schema that I missed, so within a couple days I hope to setup some tests to methodically verify that schema keyword gets converted properly.

- **Generate the `metadata?: pulumi.Input<ObjectMeta>` field** in the main <kind>Args interface. IIRC this should be the only non-custom-generated field
- **Stopped generating the top-level `metadata.name` namespace.** Previously I would generate something like:
```ts
namespace crontabs.stable.example.com {
    namespace v1 {
       interface CrontabArgs { ... }
       interface CrontabSpec { ... }
    }
    namespace v2 {
       interface CrontabArgs { ... }
       interface CrontabSpec { ... }
    }
}
```

Now I just generate the version namespaces as the top-level namespace. Developers will probably do something like `import * as crontabs from ./crontabs` anyways, so we don't really need it. It also fixes the old bug where if the namespace had pulumi in it, (e.g `stacks.pulumi.com`), then trying to declare pulumi input variables (e.g `pulumi.Input<string>`) would fail since TypeScript thinks we're referring to the pulumi namespace.

- **NestedMapSlice now returns false if the field is found, but not of type[]interface{}.** This matches the expected behavior of the other unstruct.Nested<...> methods. I can't remember exactly, but I think I first made it return true since I wanted the logic of "still return true if the value was found, but not of the proper type." I cleaned up the OpenAPI schema parsing logic a bit, so this functionality is no longer needed.

- **Removed copies of `codegen/nodejs/...` files.** I previously just directly copied the pulumi schema codegen into `crd2pulumi/nodejs/...` and made edits to it. Now I actually import from the `pulumi/pulumi` repo directly, but with a `replace` statement pointing to my local copy in `go.mod.` In my `pulumi/pulumi` [PR](https://github.com/pulumi/pulumi/pull/5084), I added some methods that directly generate the namespaces I need.

- **Add CustomResourceGenerator struct**. Contains some utility methods to quickly get commonly needed fields from the CRD (apiVersion, name, etc...) and encapsulates the logic of 1) parsing the YAML file and 2) outputting the generated code. 

- **Parse hyphened fields.** The YAML can't be directly converted to a JSON string and passed as an argument to `CustomResourceDefinition`, since some fields such as `x-kubernetes-int-or-string` are actually underscored in the pulumi SDK. I just wrote a function that traverses the unmarshalled `map[string]interface{}` YAML and renames all these fields.

### CODE REVIEW

File: provider/cmd/crd2pulumi/generate.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func NewCustomResourceGenerator(language, yamlPath, outputPath string) (CustomResourceGenerator, error) {
+	if language != DotNet && language != Go && language != NodeJS && language != Python {
+		return CustomResourceGenerator{}, errors.New("invalid language: " + language)
+	}
+
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("read file %s: %v", yamlPath, err)
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:57]: Consider importing [`"github.com/pkg/errors"`](https://github.com/pkg/errors) and wrap the error like:

```suggestion
		return CustomResourceGenerator{}, errors.Wrapf(err, "reading file: %s", yamlPath)
```

Applies throughout

File: provider/cmd/crd2pulumi/generate.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+			switch language {
+			case NodeJS:
+				extension = "ts"
+			case DotNet:
+				extension = "cs"
+			case Python:
+				extension = "py"
+			case Go:
+				extension = "go"
+			}
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:106]: import `"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"`

```suggestion
			default:
				contract.Failf("unexpected language %s", language)
			}
```

File: provider/cmd/crd2pulumi/generate.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			outputFileName := plural + "." + extension
+			return path.Join(filepath.Dir(yamlPath), outputFileName)
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:108]:
```suggestion
			return filepath.Join(filepath.Dir(yamlPath), outputFileName)
```

File: provider/cmd/crd2pulumi/generate.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		classes := gen.GenerateNodeJSClasses()
+
+		file, err := os.Create(gen.OutputPath)
+		if err != nil {
+			return fmt.Errorf("creating file at %s: %v", gen.OutputPath, err)
+		}
+		file.WriteString(types)
+		file.WriteString(classes)
+		defer file.Close()
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:157]: Nit: Move the `defer file.Close()` closer to the `os.Create` call:

```suggestion
		defer file.Close()
		file.WriteString(types)
		file.WriteString(classes)
```

File: provider/cmd/crd2pulumi/generate.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+		return nil
+	case DotNet:
+	case Python:
+	case Go:
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:161]: Let's go ahead and return errors for the languages that aren't supported yet, and add a default case with a `contract.Failf("unexpected language %s", gen.Language)`.

File: provider/cmd/crd2pulumi/generate.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	// Generates a CustomResource sub-class for a single version
+	generateResourceClass := func(sb *strings.Builder, version, kind, name, group string) {
+		argsName := version + "." + kind
+		apiVersion := fmt.Sprintf("%s/%s", group, version)
+		sb.WriteString(fmt.Sprintf("export namespace %s {\n", version))
+		sb.WriteString(fmt.Sprintf("\texport class %s extends k8s.apiextensions.CustomResource {\n", kind))
+		sb.WriteString(fmt.Sprintf("\t\tconstructor(name: string, args?: %s, opts?: pulumi.CustomResourceOptions) {\n", argsName))
+		sb.WriteString(fmt.Sprintf("\t\t\tsuper(name, { apiVersion: \"%s\", kind: \"%s\", ...args }, opts)\n", apiVersion, kind))
+		sb.WriteString(fmt.Sprintf("\t\t}\n\t}\n}\n\n"))
+	}
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/generate.go:182]: Alternatively to using `*strings.Builder`, you could do something like:

```suggestion
	generateResourceClass := func(w io.Writer, version, kind, name, group string) {
		argsName := version + "." + kind
		apiVersion := fmt.Sprintf("%s/%s", group, version)
		fmt.Fprintf(w, "export namespace %s {\n", version)
		fmt.Fprintf(w, "\texport class %s extends k8s.apiextensions.CustomResource {\n", kind)
		fmt.Fprintf(w, "\t\tconstructor(name: string, args?: %s, opts?: pulumi.CustomResourceOptions) {\n", argsName)
		fmt.Fprintf(w, "\t\t\tsuper(name, { apiVersion: \"%s\", kind: \"%s\", ...args }, opts)\n", apiVersion, kind)
		fmt.Fprintf(w, "\t\t}\n\t}\n}\n\n")
	}
```


And then:

```go
	buffer := &bytes.Buffer{}
 	for _, version := range versionNames {
 		generateResourceClass(buffer, version, kind, name, group)
 	}
 	generateDefinitionClass(buffer, kind, gen.APIVersion)

 	return buffer.String()
```

File: provider/cmd/crd2pulumi/schema.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+// Replaces the hyphens in the x-kubernetes-... fields with underscores
+func underscoreFields(schema map[string]interface{}) {
+	for field, val := range schema {
+		for _, hyphenedField := range HyphenedFields {
+			if field == hyphenedField {
+				delete(schema, field)
+				underScoredField := strings.ReplaceAll(field, "-", "_")
+				schema[underScoredField] = val
+			}
+		}
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/schema.go:131]: It'd be slightly more efficient to make `HyphenedFields` a map, e.g.:

```go
var HyphenedFields = map[string]struct{}{
	"x-kubernetes-embedded-resource":       struct{}{},
	"x-kubernetes-int-or-string":           struct{}{},
	"x-kubernetes-list-map-keys":           struct{}{},
	"x-kubernetes-list-type":               struct{}{},
	"x-kubernetes-map-type":                struct{}{},
	"x-kubernetes-preserve-unknown-fields": struct{}{},
}
```

And lookup:

```go
		if _, ok := HyphenedFields[field]; ok {
			delete(schema, field)
			underScoredField := strings.ReplaceAll(field, "-", "_")
			schema[underScoredField] = val
		}
```

File: provider/cmd/crd2pulumi/schema.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+		} else {
+			subSchemaSlice, ok := val.([]interface{})
+			if ok {
+				for _, genericSubSchema := range subSchemaSlice {
+					subSchema, ok = genericSubSchema.(map[string]interface{})
+					if ok {
+						underscoreFields(subSchema)
+					} else {
+						break
+					}
```
Comment (justinvp) [inline provider/cmd/crd2pulumi/schema.go:144]: I don't think you want to `break` here, otherwise, what if there are other values in the slice that need to be underscored?

Also, this function could be cleaned up a bit. Something like (untested):

```go
func underscoreFields(schema map[string]interface{}) {
	for field, val := range schema {
		if _, has := HyphenedFields[field]; has {
			delete(schema, field)
			underScoredField := strings.ReplaceAll(field, "-", "_")
			schema[underScoredField] = val
		}
		if subSchema, ok := val.(map[string]interface{}); ok {
			underscoreFields(subSchema)
		} else if subSchemaSlice, ok := val.([]interface{}); ok {
			for _, genericSubSchema := range subSchemaSlice {
				if subSchema, ok = genericSubSchema.(map[string]interface{}); ok {
					underscoreFields(subSchema)
				}
			}
		}
	}
}
```

File: provider/cmd/crd2pulumi/schema.go (hunk 7 above)
Comment (lblackstone) [reply provider/cmd/crd2pulumi/schema.go:131]: Even easier is using `codegen.StringSet` from pulumi/pulumi. It uses the `map[string]struct{}` construction under the hood, but the interface is clearer.

File: provider/cmd/crd2pulumi/generate.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+		APIVersion:               apiVersion,
+		Kind:                     kind,
+		Plural:                   plural,
+		Group:                    group,
+		Versions:                 versions,
+	}
+	return customResourceGenerator, nil
+}
+
+func (gen *CustomResourceGenerator) GetName() string {
```
Comment (lblackstone) [inline provider/cmd/crd2pulumi/generate.go:126]: nit: In Go, this type of method usually doesn't include a `Get` prefix, so this would be `Name()` rather than `GetName()`

File: provider/cmd/crd2pulumi/schema.go (hunk 8 above)
Comment (albert-zhong) [reply provider/cmd/crd2pulumi/schema.go:144]: I think my assumption with the OpenAPI schema was that map[string]interface{}s had to fill up the entire slice, so if we found a single interface{} that couldn't be casted to map[string]interface{}, then we could early exit. However I'm not 100% sure on this, so I'll drop the `break` statement.

File: provider/cmd/crd2pulumi/generate.go (hunk 6 above)
Comment (albert-zhong) [reply provider/cmd/crd2pulumi/generate.go:182]: Is there any performance advantage with using buffers compared to StringBuilders? Or is it just simpler since we can directly use string formatting in the `fmt.Fprintf` function?

Comment (justinvp) [reply provider/cmd/crd2pulumi/generate.go:182]: Just for the simpler `fmt.Fprintf`. I'm not sure about the perf difference between `strings.Builder` vs. `bytes.Buffer`, but I doubt it makes a measurable difference here. It looks like `strings.Builder` also implements the `io.Writer` interface, so it could be used with `fmt.Fprintf`.
